mono/mini/interp/transform.c

```diff
@@ -3260,7 +3260,7 @@ generate_code (TransformData *td, MonoMethod *method, MonoMethodHeader *header,
 			}
 			break;
 		case CEE_LDC_I4_1:
-			if (!td->is_bb_start[td->ip + 1 - td->il_code] && 
+			if (td->ip - td->il_code + 1 < td->code_size && !td->is_bb_start [td->ip + 1 - td->il_code] &&
 				(td->ip [1] == CEE_ADD || td->ip [1] == CEE_SUB) && td->sp [-1].type == STACK_TYPE_I4) {
```

Comment:
`sp [-1] ` looks suspicious too.

Reply:
This just accesses the next instruction following a LDC. I don't think is possible for such an instruction to be the end of the code stream. Where is this exactly triggered ? Aside from that, the access for is_bb_start was and is still wrong. We should never access it if we are inlining (it is only declared for the original method IL code).

Also, your change is wrong because apparently td->code_size is only declared for the orginal method, while td->ip and td->il_code are for the current method to be compiled. Hope we are not making this mistake elsewhere.

Reply:
Aha, you are right. It is inlining. Probably for all of them.

```
Thread 1 received signal SIGTRAP, Trace/breakpoint trap.
0x0000000000354f1f in generate_code (td=0x1ffefdf5f0, method=0xa5e91c8, header=0xa2809f0, generic_context=0x0, error=0x1ffefdf8d0) at interp/transform.c:3253
3253                            if (!td->is_bb_start[td->ip + 1 - td->il_code] && td->ip [1] == 0xfe && td->ip [2] == CEE_CEQ &&
(gdb) bt 9
#0  0x0000000000354f1f in generate_code (td=0x1ffefdf5f0, method=0xa5e91c8, header=0xa2809f0, generic_context=0x0, error=0x1ffefdf8d0) at interp/transform.c:3253
#1  0x000000000034eb53 in interp_inline_method (td=0x1ffefdf5f0, target_method=0xa5e91c8, header=0xa2809f0, error=0x1ffefdf8d0) at interp/transform.c:1790
#2  0x000000000035012f in interp_transform_call (td=0x1ffefdf5f0, method=0xa5e91a0, target_method=0xa5e91c8, domain=0x6075fc0, generic_context=0x0,
    is_bb_start=0xa628760 "\001", constrained_class=0x0, readonly=0, error=0x1ffefdf8d0, check_visibility=1, save_last_error=0) at interp/transform.c:2086
#3  0x0000000000356512 in generate_code (td=0x1ffefdf5f0, method=0xa5e91a0, header=0xb7728c0, generic_context=0x0, error=0x1ffefdf8d0) at interp/transform.c:3371
#4  0x000000000034eb53 in interp_inline_method (td=0x1ffefdf5f0, target_method=0xa5e91a0, header=0xb7728c0, error=0x1ffefdf8d0) at interp/transform.c:1790
#5  0x000000000035012f in interp_transform_call (td=0x1ffefdf5f0, method=0xa28ac40, target_method=0xa5e91a0, domain=0x6075fc0, generic_context=0xa28ac80,
```

Reply:
You should make sure that it works for inlining and that code_size is properly used (either kill it from TransformData and use header->code_size, or just keep it updated for inlined method to avoid further problems. Can't say I care too much whether it reports invalid IL or not.

Reply:
`is_bb_start` needs computing for inlining, right?
Only at most once per method? i.e. not for each inline attempt, or inline and regular?
Is there provision for that?

How about..I know this is lame..well, for inlining, for this specific code, we should just skip the optimization. But we should do better.

Ok, reading more code, it looks like we don't inline anything with a branch, to greatly simplify?

But I don't see what `is_bb_start`, um, it looks like it is allocated and sized once per call tree. And that IL offsets fro multiple functions are indexed in to the same array.
That seems very wrong, but I could also be reading it wrong.

I have a feeling that not inlining anything with a branch provides an easy fix though.
i.e. there isn't a need to have multiple is_bb_starts and look up from the inlined code. Just that the inlined code is considered all to occur at the point of the call to it.

So maybe just or in `inlining` before any `is_bb_start` check, or make a function out of that.

Reply:
we don't need to compute is_bb_start for inlined methods, because nobody is bb start there. Maybe this logic should be extracted in a separate method at some point.